lib/msf/core/auxiliary/ubiquiti.rb

```diff
@@ -0,0 +1,336 @@
+# -*- coding: binary -*-
+
+require 'metasploit/framework/hashes/identify'
+require 'bson'
+require 'zip'
+
+module Msf
+
+###
+#
+# This module provides methods for working with Ubiquiti equipment
+#
+###
+module Auxiliary::Ubiquiti
+  include Msf::Auxiliary::Report
+
+  def decrypt_unf(contents)
+    aes = OpenSSL::Cipher.new('aes-128-cbc')
+    aes.key = 'bcyangkmluohmars' # https://github.com/zhangyoufu/unifi-backup-decrypt/blob/master/E>
+    aes.padding = 0
+    aes.decrypt
+    aes.iv = 'ubntenterpriseap'
+    aes.update(contents)
+  end
+
+  def repair_zip(fname)
+    zip_exe = Msf::Util::Helper.which('zip')
+    if zip_exe.nil?
+      return nil
+    end
+    print_status('Attempting to repair zip file (this is normal and takes some time)')
+    temp_file = Rex::Quickfile.new("fixed_zip")
+    system("yes | #{zip_exe} -FF #{fname} --out #{temp_file.path}.zip > /dev/null")
+    if $? == 0
+      return File.read("#{temp_file.path}.zip")
+    else
+      print_error('Error fixing zip.  Attempt manually.')
+      nil
+    end
+  end
+
+
+  def bson_to_json(byte_buffer)
+    # This function takes a byte buffer (db file from Unifi read in), which is a bson string
+    # it then converts it to JSON, where it uses the 'select collection' documents
+    # as keys.  For instance a bson that contained the follow (displayed in json
+    # for ease):
+    # {"__cmd"=>"select", "collection"=>"heatmap"}
+    # {'example'=>'example'}
+    # {'example2'=>'example2'}
+    # would become:
+    # {'heatmap'=>[{'example'=>'example'}, {'example2'=>'example2'}]}
+    # this is mainly done to ease the grouping of items for easy navigation later.
+
+    buf = BSON::ByteBuffer.new(byte_buffer)
+    output = {}
+    key = ''
+
+    while buf
+      begin
+        # read the document from the buffer
+        bson = BSON::Document.from_bson(buf)
+        if bson.has_key?('__cmd')
+          key = bson['collection']
+          output[key] = []
+          next
+        end
+        output[key] << bson
+      rescue RangeError
+        break
+      end
+    end
+    output
+  end
+
+  def unifi_config_eater(thost, tport, config)
+    # This is for the Ubiquiti Unifi files.  These are typically in the backup download zip file
+    # then in the db.gz file as db.  It is a MongoDB BSON file, which can be difficult to read.
+    # https://stackoverflow.com/questions/51242412/undefined-method-read-bson-document-for-bsonmodule
+    # The BSON file is a bunch of BSON Documents chained together.  There doesn't seem to be a good
+    # way to read these files directly, so looping through loading the content seems to work with
+    # minimal repercussions.
+
+    # The file format is broken into sections by __cmd select documents as such:
+    # {"__cmd"=>"select", "collection"=>"heatmap"}
+    # we can pull the relevant section name via the collection value.
+
+    credential_data = {
```

Comment:
5 out of 7 times you use this template (via `credential_data.dup`) you set `cred[:private_type] = :password`. Why not just set it here and continue to override it as necessary when the type is a hash.

Reply:
Same as comment below.  This is a good point, but i'd like to evaluate it in brocade/juniper/cisco as well and just do the change to all the modules at once for consistency